### PR TITLE
website/blog: Becoming OpenID certified - Why standards matter

### PR DESCRIPTION
**Pull request by BeryJu**

### CODE REVIEW

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+In the end we decided to follow suit with authentik (mostly for the sake of compatibility, but also since it can make sense), and vCenter logins via authentik are now [fully supported](https://goauthentik.io/integrations/services/vmware-vcenter/) (at least from our side).
+
+### Standards in authentik
+
+Quickly touching on standards more generally in authentik; we aim to make authentik as standards-compliant as possible while retaining its feature set. For example, for SAML sources/providers, all generated responses are tested against the official SAML XML schema. The same is done for the newly added SCIM integration, where everything is equally validated.
```
Comment (tanberry) [inline website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:62]:
```suggestion
Quickly touching on standards more generally in authentik; we aim to make authentik as standards-compliant as possible while retaining its feature set. For example, for SAML sources/providers, all generated responses are tested against the official SAML XML schema. The same is done for the newly added [SCIM](https://en.wikipedia.org/wiki/System_for_Cross-domain_Identity_Management) integration, where user identity information is equally validated across all systems and domains.
```

**Review by tanberry — COMMENTED** [review]
Awesome info in here!

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+Since its introduction in [2014](<https://en.wikipedia.org/wiki/OpenID#OpenID_Connect_(OIDC)>), OIDC has been broadly adopted in the late 2010s, and now most providers support it (there's still a large list of providers that rely on pure OAuth with a custom identity layer on top, like Facebook, Twitter, Apple, but hopefully they'll migrate to OIDC eventually).
+
+The broad adoption on both sides has also increased the diversity and fragmentation, and indirectly also the diversion away from the standard.
```
[the remark this answers is not in the document]
Comment (BeryJu) [reply website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:32]: I meant both sides on both the identity providers and clients/libraries, but I think both points would be good

```suggestion
The strong adoption of OIDC by many, while other major companies have stayed with customized OAuth, as well as the large amount of available implementations of clients and identity providers, has increased the diversity and fragmentation of the IAM space, but also fueled the deviation away from a single standard
```

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+### Standards and testing
+
+One aspect where OIDC is very different to other standards is that they have a certification program, where applications can run through a (quite large) set of tests specified by the foundation itself, called [Conformance Tests](https://openid.net/certification/). This allows library authors for both clients and identity providers to ensure their implementation matches the standard, and behaves as expected. These tests are very thorough (the basic identity provider test suite contains 96 tests), which ensures that edge cases, uncommonly used features and both positive and negative flows are tested.
+
+These tests are all very well defined, they explain exactly what is tested, what the expected behavior is, and what (if any) possible issues are.
+
+For example, here you can see the overview of a test that's failed: ![Failed test overview](./failed-overview.png) and the exact request and reason why it failed: ![Failed test detail](./failed-detail.png)
+
+The exact standard definition is linked, which makes it very easy to dig in further and figure out what's supposed to happen.
```
[the remark this answers is not in the document]
Comment (BeryJu) [reply website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:42]: In the second picture, this little link 
image

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+For authentik, we use these tests to ensure that we adhere to the standard as much as possible to increase compatibility as much as possible. We've got all tests passing on the current development builds (which actually helped us a couple of bugs), and we'll be certifying the identity provider portion of authentik very soon.
```
[the remark this answers is not in the document]
Comment (BeryJu) [reply website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:44]: will be certifying means getting an OpenID certification from the foundation for the OpenID Identity Provider component of authentik

And while I would love to run these tests automatically before every release, the tests require a lot of manual work so that's sadly not yet possible

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+![Successful tests](./summary-green.png)
+
+### Deviation of standards
+
+It is sadly also quite common for clients to deviate from these standards, often requiring special behaviour that is specific to one or more implementations. One example that we had this issue with in the past is VMware's vCenter.
+
+They introduced "Identity federation" in [Version 7](https://blogs.vmware.com/vsphere/2020/04/vsphere-7-new-generation-vsphere.html), but explicitly only supported Microsoft's ADFS identity provider. However under the hood, this was using OIDC, which ADFS has supported for a while.
```
[the remark this answers is not in the document]
Comment (BeryJu) [reply website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:52]:
```suggestion
VMWare introduced "Identity federation" in [Version 7](https://blogs.vmware.com/vsphere/2020/04/vsphere-7-new-generation-vsphere.html), but explicitly only supported Microsoft's ADFS identity provider. However, under the hood vCenter was itself actually using OIDC, which ADFS has supported for a while.
```

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+I'll spare you the details, but after a lot of digging through logs and figuring out what vCenter actually expects and attempts to do, I found it expects the `access_token` to be an encoded [JWT](https://jwt.io/) (stay tuned for our upcoming blog post about how JWT took the identity world by storm). Expecting an encoded JWT is not part of the [OpenID standard](https://openid.net/specs/openid-connect-core-1_0.html), so it somewhat made sense that they only advertise ADFS compatibility. However as we were finding out, vCenter was not the only applications that had this requirement. Researching further, it seemed like this had become sort of a "quasi-standard", as many identity providers were behaving this way.
+
+In the end we decided to follow suit with authentik (mostly for the sake of compatibility, but also since it can make sense), and vCenter logins via authentik are now [fully supported](https://goauthentik.io/integrations/services/vmware-vcenter/) (at least from our side).
```
[the remark this answers is not in the document]
Comment (BeryJu) [reply website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:58]: authentik used to use JWTs before this as well, this specifically is to say we followed suit with using JWTs within the `access_token`

File: website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md (hunk 1 above)
Comment (BeryJu) [reply website/blog/2023-03-07-becoming-openid-certified-why-standards-matter/index.md:62]: Oh yeah that is pretty unclear; yeah I wanted to say to stay as close as possible to OIDC standards (or rather standards in general) while also retaining our feature set

I think for this post it makes more sense to be OIDC-compliant since that's the topic, but the general goal is to be compliant with all standards authentik supports
